### PR TITLE
feat: readiness check

### LINKED ISSUE #1687: Ping endpoint should check Redis connection
## Expected Behavior

ping endpoint should check Redis connection

## Current Behavior

ping endpoint always returns 'OK' message 

## Possible Solution
using something like validating Redis configuration on startup

## Context
when running oauth2-proxy on K8S environment with liveness/readiness probe, if Redis server down oauth2-proxy still in the ready state because of ping endpoint always returns 'OK'.

## Your Environment

- Version used: 7.3.0

### PR DESCRIPTION
**Pull request by kobim**
Adds an in-depth health check to verify the underlying components (e.g., Redis) are connected.

## Description

If the proxy needs an underlying store (e.g., Redis), it should be up & running. This addition adds the option to perform a deeper check of the underlying session store.

With the Redis store, it is achieved by using the [PING command](https://redis.io/commands/ping/).

## Motivation and Context

When running the `oauth2-proxy`, we should verify if it's functioning. If the Redis server is down, the proxy will still serve traffic as it doesn't verify the server is available.

Fixes #1687 

## How Has This Been Tested?

- Unit tests
- Running the proxy with a redis session store. Once the proxy runs, call the health endpoint when the redis is still running and when it is killed.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: docs/docs/features/endpoints.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ title: Endpoints
 OAuth2 Proxy responds directly to the following endpoints. All other endpoints will be proxied upstream when authenticated. The `/oauth2` prefix can be changed with the `--proxy-prefix` config variable.
 
 - /robots.txt - returns a 200 OK response that disallows all User-agents from all paths; see [robotstxt.org](http://www.robotstxt.org/) for more info
-- /ping - returns a 200 OK response, which is intended for use with health checks
+- /ping - returns a 200 OK response, which is intended for use with health checks. If the `X-Healthcheck: full` header is present, it performs a deeper check for underlying connections (e.g., Redis store)
```
Comment (JoelSpeed) [inline docs/docs/features/endpoints.md:9]: Would it be better to have a separate endpoint? Can for example a Kubernetes readiness check configure headers to send?

**Review by JoelSpeed — COMMENTED** [review]
Looks pretty good! Got a question about headed vs path though

File: docs/docs/features/endpoints.md (hunk 1 above)
Comment (kobim) [reply docs/docs/features/endpoints.md:9]: I started with a different path and converted to using the headers instead.
I had the _manifests_ in mind (https://github.com/oauth2-proxy/manifests/pull/113) and was looking for a suitable upgrade strategy (how people won't get confused and deploy a new manifest with an old version that doesn't support the new path).

That said, I have no hard feelings about using headers vs. different paths (e.g., `/ping` and `/ready`). Maybe I could use [Helm's `semverCompare`](https://helm.sh/docs/chart_template_guide/function_list/#semvercompare) to make this distinction instead.

What do you think?

Comment (JoelSpeed) [reply docs/docs/features/endpoints.md:9]: Is the version of the helm chart not tied to the version of the application within it? My understanding is that if we add the new flags to the helm chart, they will only be done so once a new version of OAuth2 proxy is released

I would lean towards a separate endpoint as I think in general that has more compatibility

File: pkg/middleware/readynesscheck.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/justinas/alice"
+)
+
+type Verifiable interface {
```
Comment (JoelSpeed) [inline pkg/middleware/readynesscheck.go:11]: Please add godoc to each function, interface etc to explain what they are for

**Review by JoelSpeed — COMMENTED** [review]
Looks good, please make sure every function has some appropriate documentation and I'll get this merged

**Comment by pierluigilenoci** [comment]
@JoelSpeed could you please take a look?

**Comment by pierluigilenoci** [comment]
@JoelSpeed could you please take a look?
@kobim could you please resolve the conflicts?

**Review by JoelSpeed — APPROVED** [review]
LGTM thanks

**Comment by sathieu** [comment]
Is there a new release planned, including this PR?

**Pull request merged**